docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,256 @@
+
+Receiving broadcasts from other evidence reactors works in the same manner.
+
+When it comes to prevoting and precomitting a proposal that contains evidence, the full node will once again
+call upon the evidence pool to verify the evidence using `CheckEvidence(ev []Evidence)`:
+
+This loops through all the evidence first checking that nothing has been duplicated then:
+
+runs `fastCheck(ev evidence)` which works similar to `Has` but with `LightClientAttackEvidence` if it has the
+same hash it then checks to see that the validators it has are all signers in the commit of the conflicting header. If it doesn't pass fast check (because it hasn't seen the evidence before) then it will have to verify the evidence.
+
+runs `Verify(ev Evidence)` - Note: this also saves the evidence to the db as mentioned before.
+
+
+The final part of the lifecycle is when the block is committed and the `BlockExecutor` tries to apply state. Instead of trying to form the information necessary to produce the ABCI evidence it gets the evpool to create the evidence in the update step as follows.
+
+`Update(Block, State) []abci.Evidence`
+
+```go
+abciResponses.BeginBlock.ByzantineValidators = evpool.Update(block, state)
+```
+
+Update does the following:
+- increments state which keeps track of both the current time and height used for measuring expiry
+- marks evidence as committed and saves to db. This prevents validators from proposing committed evidence in the future
+  Note: the db just saves the height and the hash. There is no need to save the entire committed evidence
+- forms abci evidence:
+```go
+for _, val := range evInfo.Validators {
+  abciEv = append(abciEv, &abci.Evidence{
+    Type: evType,
+    Validator: val,
+    Height: evInfo.ev.Height(),
+    Time: evInfo.time,
+    TotalVotingPower: evInfo.totalVotingPower
+  })
+}
+```
+and compiles it all together (note for `DuplicateVoteEvidence` the validators array size is 1)
+- removes expired evidence from both pending and committed
+
+The abci evidence is then sent via the `BlockExecutor` to the application.
+
+To summarize, we can see the lifecycle of evidence as such:
+
+![evidence_lifecycle](../imgs/evidence_lifecycle.png)
+
+Evidence is first detected and created in the light client and consensus engine. It is verified and stored as `EvidenceInfo` and gossiped to the evidence pools in other nodes. The consensus engine later communicates with the evidence pool to either retrieve evidence to be put into a block, or verify the evidence the consensus engine has retrieved in a block. Lastly when a block is added to the chain, the block executor sends the committed evidence back to the evidence pool so a pointer to the evidence can be stored in the evidence pool and it can update it's height and time. Finally, it turns the committed evidence into ABCI evidence and through the block executor passes the evidence to the application so the application can handle it.
```

Comment:
This may be context that most reviewers would already have, but I'm still confused about what exactly the evidence pool is. I think this ADR would be even better with an extra paragraph on what the evidence pool does/what logic it has/what it handles and persists.

Reply:
I will try add more of an introductory paragraph at the start of the detailed design. Essentially the entire ADR is dedicated towards the evidence pool and the duties it performs.